### DISCUSSION TITLE
Migrating from jinja to scriban

### DISCUSSION THREAD

**Discussion by tevoinea** [microsoft/onefuzz#2491] [Ideas]
We currently accept notification templates using the jinja templating language. Jinja isn't supported in C# so we must migrate to a new templating engine. Scriban was chosen because it's inspired by jinja so the templates are very similar.

### The current state

Notification templates are currently being processed and rendered by the python code. Once #2486 merges, we'll be able to adapt our existing jinja templates to be compatible with scriban. The template is not updated in the database, only in memory. This is intended to be a 'compatibility mode' until the migration is complete.

> **Note:** The adapter is capable of converting all of the _current_ jinja templates in our prod instance. It is not capable of adapting all possible jinja templates.

### The desired state

We only want to accept scriban templates. We only store scriban templates in our database.

### How do we get there?

1. Enable the C# version of `QueueFileChanges`.
    1. Internally, this should be a no-op because we know the adapter is capable of converting all of our existing templates to a scriban compatible format.
    1. This also unblocks the last step of C# migration.
1. Create a scriban template verification endpoint.
1. Notify users of the upcoming breaking change, instruct them to update notification templates using the new verification endpoint. It could be nice to have a cli command for this.
1. Select a date for the breaking change. The breaking change will:
    1. Enforce all new notification templates must pass scriban template verification before creation.
    3. Enable the C# implementation of `QueueFileChanges` as the default when deploying OneFuzz.
1. After the breaking change date, do a final 'courtesy' migration of any jinja templates in the database -- this time updating the entries.
1. Remove any C# code with references to jinja.

### Potential issues

* The adapter cannot convert all possible jinja templates to scriban so switching to the C# implementation can be a breaking change for our open source users.

**Comment by stishkin** [comment]
`Create a scriban template verification endpoint.` - you mean http endpoint where user submits their converted template and they get pass or fail ?

**Reply by tevoinea** [reply]
Exactly that. I was thinking the user would submit a template and we can render it for them using demo data so not only would they know if it's a valid scriban template but also validate that it looks like they expect.

> It could be nice to have a cli command for this.

Could be something like `onefuzz debug notification_template @./myconfig.json` that hits the verification endpoint.

**Comment by chkeita** [comment]
We should also start versioning the templates at this point. And we can assume from this point that a template with no version is a jinja template. We could optionally add a language field to the template too.
